File: 백준/Gold/18427. 함께 블록 쌓기/함께 블록 쌓기.java
import java.util.*;
import java.io.*;

class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int h = Integer.parseInt(st.nextToken());
		
		int[] dp = new int[1001];
		dp[0] = 1;
		int mod = 10007;
		for(int i=0;i<n;i++) {
			st = new StringTokenizer(br.readLine());
			int[] ndp = new int[1001];
			for(int j=0;j<=h;j++) ndp[j] = dp[j];
			while(st.hasMoreTokens()) {
				int a = Integer.parseInt(st.nextToken());
				for(int j=a;j<=h;j++) ndp[j] = (ndp[j] + dp[j-a]) % mod;
			}
			for(int j=0;j<=h;j++) dp[j] = ndp[j];
		}
		bw.write(dp[h] + "\n");
		
		bw.flush();
		bw.close();
	}
}

File: 백준/Gold/18427. 함께 블록 쌓기/함께 블록 쌓기.py
N, M, H = map(int, input().split())
a = []
for i in range(N):
    a.append(list(map(int, input().split())))

MOD = 10007
dp = [ 0 for _ in range(H+1) ]
dp[0] = 1

for i in range(N):
    ndp = [ 0 for _ in range(H+1) ]
    a[i].append(0)
    for j in a[i]:
        for k in range(0, H-j+1):
            ndp[k+j] += dp[k]
            ndp[k+j] %= MOD
    dp = ndp

print(dp[H])